[bugfix] TensorList args in functionalization pass

bdhirsh opened the pull request:
At the time that I wrote the pass, I thought that `c10::TensorList` and `c10::List<Tensor>` were the same thing. But it looks like a `TensorList` is actually an `ArrayRef<Tensor>`. This led to a nasty bug when I tried to add conditional functionalization to `block_diag`, where in the boxed kernel, I would:

(1) unwrap the first `IValue` by calling `.toTensorList()` (this actually returns a `List<Tensor>`, not a `TensorList`).
(2) call `TensorList to_functional_tensor(List<Tensor>)` to get out a `TensorList` with the functionalized tensors
(3) wrap that back into an `IValue` and put in on the stack.

Somewhere in that sequence of operations, something bad happens and we segfault. Fixing up the signature of `to_functional_tensor` to be `List<Tensor> to_functional_tensor(List<Tensor>)` fixes the bug. I have a feeling that there's a latent TensorList-related bug in the boxing/unboxing logic that made this worse, but I'm okay to stick with my narrow fix for now.

Additionally tested by running `pytest test/test_ops.py test/test_vmap.py -v -k block_diag` on top of this PR: https://github.com/pytorch/functorch/pull/235

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#68395 [bugfix] TensorList args in functionalization pass**

Differential Revision: [D32448258](https://our.internmc.facebook.com/intern/diff/D32448258)

bdhirsh committed:
[bugfix] TensorList args in functionalization pass

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32448258).

zou3519 commented on test/test_functionalization.py:
```diff
@@ -85,6 +85,25 @@ def f(x):
+        logs = self.get_logs(f, torch.ones(2, 2))
+        # Only seeing copy_() calls in the logs are actually expected:
+        # - block_diag is a CompositeImplicitAutograd op, implemented in terms of copy_() and a few other ops.
+        # - copy_() doesn't have an out-of-place variant, so the pass leaves it alone
```
Wait, so when do we turn copy_() into .to(...).expand_as(...) ? Is this a bug?

bdhirsh replied:
This is actually expected, but I just realized that it will pose some problems for LazyTensor/XLA/Vulkan later.

This happens because `block_diag` doesn't decompose during the functionalization pass: instead, `block_diag` will hit the functionalization boxed fallback, redispatch, and decompose **underneath** the pass (so its base ops will not be functionalized).

The easiest solution is to just say that all `CompositeImplicitAutograd` ops should decompose during functionalization. I remember trying that initial and running into some problems - I forget exactly what they were, but one I can think of is that this would prevent XLA/LTC/Vulkan from providing their own custom implementations of composite ops.

We could actually maybe get around that in a way similar to how we deal with the `Autograd` key right now - by making functionalization per backend concept, registering everything to an alias `Functionalization` key, and letting backends override composite kernels by registering their own implementations to per-backend keys.

Another simpler option would just be to do the same thing that we're doing for functorch vmap in core: maintain a list of "problematic" composite that get explicit functionalization kernels.

I think this probably doesn't need to be addressed for now, but will definitely come up next year when trying to integrate the pass with LazyTensor.

But the downside is that backends can't "recover" the composite op after it's decomposed, if they want to provide a custom kernel. I think we could actually get around this similar to how

zou3519 replied:
> This happens because block_diag doesn't decompose during the functionalization pass: instead, block_diag will hit the functionalization boxed fallback, redispatch, and decompose underneath the pass (so its base ops will not be functionalized).

Oh that makes a lot of sense, thank you for the clarification.

> But the downside is that backends can't "recover" the composite op after it's decomposed, if they want to provide a custom kernel. I think we could actually get around this similar to how

This is sounding very similar to the vmap(vjp issue we were discussing earlier today

bdhirsh replied:
It definitely does 😬 . "An op has (potentially multiple) decompositions, that should be toggleable/customizable".

Alias keys are one tool that we have to solve this type of problem: We register the "vanilla" decomposition to the alias key, and allow backends to customize it by registering something to their own per-backend key instead. That doesn't work for functorch though, since functorch isn't a "backend". Maybe there's something there though.

zou3519 reviewed (approved):
Changes LGTM, the behavior in the test is a bit surprising but I'm probably missing something

bdhirsh committed:
Update on "[bugfix] TensorList args in functionalization pass"

At the time that I wrote the pass, I thought that `c10::TensorList` and `c10::List<Tensor>` were the same thing. But it looks like a `TensorList` is actually an `ArrayRef<Tensor>`. This led to a nasty bug when I tried to add conditional functionalization to `block_diag`, where in the boxed kernel, I would:

(1) unwrap the first `IValue` by calling `.toTensorList()` (this actually returns a `List<Tensor>`, not a `TensorList`).
(2) call `TensorList to_functional_tensor(List<Tensor>)` to get out a `TensorList` with the functionalized tensors
(3) wrap that back into an `IValue` and put in on the stack.

Somewhere in that sequence of operations, something bad happens and we segfault. Fixing up the signature of `to_functional_tensor` to be `List<Tensor> to_functional_tensor(List<Tensor>)` fixes the bug. I have a feeling that there's a latent TensorList-related bug in the boxing/unboxing logic that made this worse, but I'm okay to stick with my narrow fix for now.

Additionally tested by running `pytest test/test_ops.py test/test_vmap.py -v -k block_diag` on top of this PR: https://github.com/pytorch/functorch/pull/235


Differential Revision: [D32448258](https://our.internmc.facebook.com/intern/diff/D32448258)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32448258).
